pkg/controller/cronjob/cronjob_controller.go

```diff
@@ -148,6 +149,23 @@ func SyncOne(sj batch.CronJob, js []batch.Job, now time.Time, jc jobControlInter
 			recorder.Eventf(&sj, v1.EventTypeNormal, "SawCompletedJob", "Saw completed job: %v", j.Name)
 		}
 	}
+
+	// Remove any job reference from the active list if the corresponding job does not exist any more.
+	// Otherwise, the cronjob may be stuck in active mode forever even though there is no matching
+	// job running.
+
+	childrenJobs := make(map[types.UID]bool)
+	for _, childJob := range js {
+		childrenJobs[childJob.ObjectMeta.UID] = true
+	}
+	for _, j := range sj.Status.Active {
+		found := childrenJobs[j.UID]
+		if !found {
+			recorder.Eventf(&sj, v1.EventTypeNormal, "MissingJob", "Active job went missing: %v", j.Name)
+			deleteFromActiveList(&sj, j.UID)
+		}
+	}
+
```

Comment:
I think you could move this to the above loop which already goes through `js` and calls `inActiveList` which does exactly what you need. You should split the `if !found && !IsJobFinished(&j)` condition into two, one dealing with the current situation and the other with what you want, iow:
```go
if !found {
    if !IsJobFinished(&j) {
        recorder.Eventf(&sj, v1.EventTypeWarning, "UnexpectedJob", "Saw a job that the controller did not create or forgot: %v", j.Name)
    } else {
        recorder.Eventf(&sj, v1.EventTypeNormal, "MissingJob", "Active job went missing: %v", j.Name)
        deleteFromActiveList(&sj, j.UID)
    }
}
```